### PR TITLE
Add pvl instruction tests

### PR DESCRIPTION
**Pull request by mlsteele**
Adds some tests for each instruction. These tests should cover normal operation of pvl (both succeeding and failing instructions). They do not cover the darker corners of malformed pvl, running dns proofs, nor malformed top-level pvl.

Other goodies:
- add mocking mode to `StubAPIEngine`
- fixed crash on negative indices in selectors
- added `func pyindex` so that negative indices act as specified

r? @maxtaco @oconnor663

### CODE REVIEW

**Comment by maxtaco** [comment]
@mlsteele see C/I failures, I think they're real failures introduced with this change.

File: go/libkb/api_stub.go [hunk 1]
```diff
@@ -5,31 +5,146 @@
...
+
+	resjson, okjson := e.expectingJSON[arg.Endpoint]
+	reshtml, okhtml := e.expectingHTML[arg.Endpoint]
+	restext, oktext := e.expectingText[arg.Endpoint]
+
+	ok := (okjson || (restype != XAPIResJSON)) &&
+		(okhtml || (restype != XAPIResHTML)) &&
+		(oktext || (restype != XAPIResText))
+	if !ok {
+		fmt.Printf("endpoint: %v\n", arg.Endpoint)
```
Comment (maxtaco) [inline go/libkb/api_stub.go]: is it possible to use logging for this?

File: go/libkb/api_stub.go [hunk 2]
```diff
@@ -5,31 +5,146 @@
...
+		mocking:           false,
+		expectingJSON:     make(map[string]*ExternalAPIRes),
+		expectingHTML:     make(map[string]*ExternalHTMLRes),
+		expectingText:     make(map[string]*ExternalTextRes),
+		calls:             make([]stubAPIEngineCallRecord, 0),
 	}
 }
 
 func (e *StubAPIEngine) Get(arg APIArg) (res *ExternalAPIRes, err error) {
+	if e.mocking {
```
Comment (maxtaco) [inline go/libkb/api_stub.go]: in what case would this StubAPI be linked to from `G` but we would want mocking off?

File: go/libkb/api_stub.go [hunk 3]
```diff
@@ -5,31 +5,146 @@
...
+		}
+		return res, nil
+	}
 	return e.ExternalAPIEngine.GetHTML(arg)
 }
 
 func (e *StubAPIEngine) GetText(arg APIArg) (*ExternalTextRes, error) {
-	if body, ok := apiStubs[arg.Endpoint]; ok {
-		return &ExternalTextRes{Body: body, HTTPStatus: 200}, nil
+	if e.mocking {
```
Comment (maxtaco) [inline go/libkb/api_stub.go]: the CI failure is that the old mode of behavior broke here.

File: go/libkb/api_stub.go [hunk 4]
```diff
@@ -5,31 +5,146 @@
...
+}
+
+func (e *StubAPIEngine) AssertCalledOnceWith(kind XAPIResType, endpoint string) error {
+	if len(e.calls) == 0 {
+		return fmt.Errorf("stub api not called")
+	}
+	if len(e.calls) > 1 {
+		return fmt.Errorf("stub api called more than once")
+	}
+	call := e.calls[0]
```
Comment (maxtaco) [inline go/libkb/api_stub.go]: do you need to reset after this?

File: go/libkb/api_stub.go [hunk 5]
```diff
@@ -5,31 +5,146 @@
...
+type stubAPIEngineCallRecord struct {
+	kind     XAPIResType
+	endpoint string
 }
 
 func NewStubAPIEngine() *StubAPIEngine {
 	return &StubAPIEngine{
 		ExternalAPIEngine: &ExternalAPIEngine{BaseAPIEngine{clients: make(map[int]*Client)}},
+		mocking:           false,
+		expectingJSON:     make(map[string]*ExternalAPIRes),
```
Comment (maxtaco) [inline go/libkb/api_stub.go]: would id make more sense to have one map that maps a string a struct that has 3 fields?

File: go/pvl/interp.go [hunk 6]
```diff
@@ -95,22 +95,42 @@ var Steps = map[CommandName]Step{
 	CmdTransformURL:        stepTransformURL,
 }
 
+// ProofInfo contains all the data about a proof PVL needs to check it.
+// It can be derived from a RemoteProofChainLink and SigHint.
```
Comment (maxtaco) [inline go/pvl/interp.go]: Ok, great!

File: go/libkb/api_stub.go [hunk 7]
```diff
@@ -5,31 +5,146 @@
...
+		if err != nil {
+			return nil, err
+		}
+		return res, nil
+	}
 	return e.ExternalAPIEngine.GetHTML(arg)
 }
 
 func (e *StubAPIEngine) GetText(arg APIArg) (*ExternalTextRes, error) {
-	if body, ok := apiStubs[arg.Endpoint]; ok {
```
Comment (maxtaco) [inline go/libkb/api_stub.go]: you dropped this!

File: go/pvl/interp_test.go [hunk 8]
```diff
@@ -1,4 +1,647 @@
...
+]`},
+		service:    keybase1.ProofType_HACKERNEWS,
+		restype:    libkb.XAPIResText,
+		restext:    "1234",
+		shouldwork: false,
+	}, {
+		name:      "AssertRegexMatch-content-ok",
+		proofinfo: info1,
+		prepvl: map[keybase1.ProofType]string{
+			keybase1.ProofType_REDDIT: `[
```
Comment (maxtaco) [inline go/pvl/interp_test.go]: should this be a twitter proof, and not a reddit proof?

Comment (mlsteele) [inline go/pvl/interp_test.go]: These are all fake anyway. I tried to vary which ProofType is used just to encourage bugs in any future service-specific code to show up.

**Comment by mlsteele** [comment]
- restored libkb/api_stub.go (should fix test failures)
- added pvl/api_stub.go
- got rid of the `mocking` flag (always on for `pvl/api_stub.go`)
- removed accidental printfs
- made stub api expectations a map to a triplet-struct

**Comment by maxtaco** [comment]
I'm a 👍, great work @mlsteele !!!!

**Pull request merged**